designs/replicated-pv/mayastor/rwx-kubevirt.md

```diff
@@ -0,0 +1,261 @@
+    nvme-of-initiator-1 --> |nvme-of| io-engine-3
+    nvme-of-initiator-2 --> |nvme-of| io-engine-3
+    Core --> |Internal Api| io-engine-3
+```
+
+<br>
+
+Optionally, we may modify the csi-node to take the reservations when connecting, though this would have to be carefully considered and I'm not planning on doing it for this.
+
+Here's how the workflow would look like:
+
+<br>
+
+```mermaid
+sequenceDiagram
+  Actor CO
+  participant CSI Controller
+
+  participant PublicApi
+  participant Core
+  participant IoEngine-1
+
+  CO ->>+ CSI Controller: Publish Volume to Node-1
+  CSI Controller ->> PublicApi: Publish Volume to Node-1
+  PublicApi ->>+ Core: Publish Volume to Node-1
+  alt First publishes imples volume target creation
+  Core ->> IoEngine-1: Create Volume Target (Nexus)
+  IoEngine-1 -->> Core: 
+  Core ->> IoEngine-1: Add Node-1 as allowed client
+  IoEngine-1 -->> Core: 
+  end
+  Core -->>- PublicApi: 
+  PublicApi -->> CSI Controller: 
+  CSI Controller -->>- CO: 
+
+  CO ->>+ CSI Controller: Publish Volume to Node-2
+  Note right of CSI Controller: Ensure<br> mode is MultiNodeMultiWriter<br>and access type is Block
+  CSI Controller ->> PublicApi: Publish Volume to Node-2
+  PublicApi ->>+ Core: Publish Volume to Node-2
+  alt Node-1 and Node-2 are both allowed
+  Core ->> IoEngine-1: Add Node-2 as allowed client
+  IoEngine-1 -->> Core: 
+  end
+  Core -->>- PublicApi: 
+  PublicApi -->> CSI Controller: 
+  CSI Controller -->>- CO: 
+```
+
+<br>
+
+#### Public OpenAPI
+
+No changes required.
+
+#### Internal gRPC
+
+No changes required.
+
+#### Kubectl Plugin
+
+No changes required.
+
+<br>
+
+### Test Plan
+
+Testing this features ofcourse requires exercising multi-attach scenarios where more than 1 workload is scheduled for the block volume, on different nodes.
+We also need to ensure we reject multi-attach for filesystem volumes. \
+The HA feature also needs to be tested, ensuring there's no flapping of the volume target, should multiple initiators attempt to trigger a failover.
+
+<br>
+
+#### Behaviour specification
+
+<br>
+
+### Risks and Mitigations
+
+To ensure high-availability, when the connection from the initiator to the target becomes failed we trigger failover of the volume target to other nodes.
+With multi-attach we now have multiple initiator-target connections, which may request for failover all at once, or in short sequence! This could wreak avoc in the system, leading into a self-feeding loop of failovers.
+
+<br>
+
+#### Mitigations
+
+In order to mitigate this, we must ensure the failover logic is hardened and resistive to short-succesion failovers, as well as prescriptive forcing the initiator to attempt reconnection against "new-existing" targets. By this I mean, if a previous failover creates a new target, then the next failover requests shoulds try to connect this "new-existing" target, rather than causing an even newewer target to be created, which would then break the first failover!
```

Comment:
```suggestion
In order to mitigate this, we must ensure the failover logic is hardened and resistive to short-succesion failovers, as well as prescriptive forcing the initiator to attempt reconnection against "new-existing" targets. By this I mean, if a previous failover creates a new target, then the next failover requests should try to connect this "new-existing" target, rather than causing an even newer target to be created, which would then break the first failover!
```